caffe2/onnx/onnx_exporter.cc

```diff
@@ -0,0 +1,585 @@
+      };
+  return kSpecialOperators;
+}
+
+void OnnxExporter::CopyCaffe2ArgToOnnxAttr(
+    AttributeProto* attr,
+    const std::string& op_type,
+    const caffe2::Argument& arg) {
+  std::string name;
+  if (get_per_op_renamed_attrs().count(op_type)) {
+    name = caffe2::get_default(
+        get_per_op_renamed_attrs().at(op_type), arg.name(), arg.name());
+  } else {
+    name = caffe2::get_default(get_renamed_attrs(), arg.name(), arg.name());
+  }
+  attr->set_name(name);
+
+  if (arg.has_f()) {
+    attr->set_f(arg.f());
+    attr->set_type(AttributeProto::FLOAT);
+  } else if (arg.has_i()) {
+    attr->set_i(arg.i());
+    attr->set_type(AttributeProto::INT);
+  } else if (arg.has_s()) {
+    attr->set_s(arg.s());
+    attr->set_type(AttributeProto::STRING);
+  } else if (arg.floats_size()) {
+    attr->mutable_floats()->CopyFrom(arg.floats());
+    attr->set_type(AttributeProto::STRINGS);
+  } else if (arg.ints_size()) {
+    attr->mutable_ints()->CopyFrom(arg.ints());
+    attr->set_type(AttributeProto::INTS);
+  } else if (arg.strings_size()) {
+    attr->mutable_strings()->CopyFrom(arg.strings());
+    attr->set_type(AttributeProto::STRINGS);
+  } else {
+    CAFFE_THROW(
+        caffe2::MakeString("Unsupported Caffe2 argument: ", arg.name()));
+  }
+}
+
+bool OnnxExporter::IsBlaskListed(const caffe2::Argument& arg) {
+  const static std::unordered_map<std::string, std::unordered_set<std::string>>
+      kBlackListString = {{"order", {"NCHW"}}};
+  const static std::unordered_map<std::string, std::unordered_set<int64_t>>
+      kBlackListInt = {{"cudnn_exhaustive_search", {0, 1}},
+                       {"use_cudnn", {0, 1}}};
+
+  if (arg.has_i()) {
+    const auto it = kBlackListInt.find(arg.name());
+    if (it != kBlackListInt.end()) {
+      return it->second.count(arg.i());
+    }
+  } else if (arg.has_s()) {
+    const auto it = kBlackListString.find(arg.name());
+    if (it != kBlackListString.end()) {
+      return it->second.count(arg.s());
+    }
+  }
+
+  return false;
+}
+
+ConvertedResult OnnxExporter::Caffe2OpToOnnxNodes(
+    const caffe2::OperatorDef& def,
+    const std::unordered_map<std::string, caffe2::TensorShape>& shapes) {
+  std::string type = def.type();
+  const auto& renamed_op_lut = get_renamed_operators();
+  const auto it = renamed_op_lut.find(type);
+  if (it != renamed_op_lut.end()) {
+    type = it->second;
+  }
+  if (get_special_operators().count(type)) {
+    return (this->*get_special_operators().at(type))(def, shapes);
+  } else {
+    return CommonCaffe2OpToOnnxNodes(def);
+  }
+}
+
+ConvertedResult OnnxExporter::CommonCaffe2OpToOnnxNodes(
+    const caffe2::OperatorDef& def) {
+  ConvertedResult result;
+  auto& nodes = result.first;
+  nodes.emplace_back();
+  NodeProto& node = nodes.back();
+  node.set_name(def.name());
+  node.set_op_type(
+      caffe2::get_default(get_renamed_operators(), def.type(), def.type()));
+  for (const auto& i : def.input()) {
+    node.add_input(i);
```

Comment:
IIRC, Caffe2 operators can overwrite previous outputs, but ONNX operators can not. You may need extra logic to handle these situations.

Reply:
Yes. That's why I added the comment that the input NetDef should be SSA rewritten form already. Currently, the SSA rewriting part is in Python, I plan to move that into C++.